### ISSUE TITLE
while_let_loop false positive

### ISSUE THREAD

**Issue by crumblingstatue** [Manishearth/rust-clippy#1693]
```rust
fn main() {
    let mut lines = "foo".lines().peekable();
    // Gather remaining non-empty lines into a comment field.
    let mut comment = String::new();
    loop {
        if let Some(line) = lines.peek() {
            if line.is_empty() {
                break;
            } else {
                comment += line;
            }
        } else {
            break;
        }
        // Consume
        lines.next();
    }
}
```

This can't be trivially transformed into `while let`, since `lines.next()` is called unconditionally outside of the `if let` statement.

**Comment by ayosec** [comment]
I have the same issue. My code:

```rust
let mut level = 1;
loop {
    match blocks.peek().map(|n| is_tb(&n.data.borrow().value)) {
        Some(true) => level += 1,
        _ => break,
    }
    blocks.next();
}
```

Clippy suggest to change it to 

```rust
while let Some(true) = blocks.peek().map(|n| is_tb(&n.data.borrow().value)) {
    ...
}
```

*Maybe*, clippy should ignore loops when `peek()` is used.

---

Full message:

```
warning: this loop could be written as a `while let` loop
   --> src/parser/blocks.rs:248:5
    |
248 | /     loop {
249 | |         match blocks.peek().map(|n| is_tb(&n.data.borrow().value)) {
250 | |             Some(true) => level += 1,
251 | |             _ => break,
252 | |         }
253 | |         blocks.next();
254 | |     }
    | |_____^ help: try `while let Some(true) = blocks.peek().map(|n| is_tb(&n.data.borrow().value)) { .. }`
    |
    = note: #[warn(while_let_loop)] on by default
    = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#while_let_loop

```

**Comment by montrivo** [comment]
> This can't be trivially transformed into while let, since lines.next() is called unconditionally outside of the if let statement.

In this case the `lines.next()`, or any code following the `if let _ {} else {break}` expression is reachable only if the pattern matches. The lint seems valid since it can be rewritten as
```rust
fn main() {
    let mut lines = "foo".lines().peekable();
    // Gather remaining non-empty lines into a comment field.
    let mut comment = String::new();
 
    while let Some(line) = lines.peek() {
        if line.is_empty() {
            break;
        } else {
            comment += line;
        }
        // Consume
        lines.next();
    }
}
```

Maybe the suggestion could highlight the necessary code mode.

Same case for second example:
```rust
while let Some(true) = blocks.peek().map(|n| is_tb(&n.data.borrow().value)) {
    level += 1;
    blocks.next();
}
```

**Comment by alexheretic** [comment]
The above code doesn't compile though. This is still a false positive as while-peek-next does not compile.

However, while-peek-next _does_ work with NLL. So maybe this lint should be disabled until nll lands in nightly default-enabled.


![while-let-loop](https://user-images.githubusercontent.com/2331607/42085038-3cca0520-7b87-11e8-95a4-8ac65940dce2.jpg)

![while-let-loop-nll](https://user-images.githubusercontent.com/2331607/42085091-5e26895a-7b87-11e8-9e0f-408b4921b0ab.jpg)
